controllers/addon_controller.go

```diff
@@ -173,71 +182,112 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
 	return ret, procErr
 }
 
-// SetupWithManager is called to setup manager and watchers
-func (r *AddonReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	log := r.Log
-	managedNS := "addon-manager-system"
+func New(mgr manager.Manager, stopChan <-chan struct{}) (controller.Controller, error) {
+	r := &AddonReconciler{
+		Client:          mgr.GetClient(),
+		Log:             ctrl.Log.WithName(controllerName),
+		Scheme:          mgr.GetScheme(),
+		versionCache:    addon.NewAddonVersionCacheClient(),
+		dynClient:       dynamic.NewForConfigOrDie(mgr.GetConfig()),
+		generatedClient: kubernetes.NewForConfigOrDie(mgr.GetConfig()),
+		recorder:        mgr.GetEventRecorderFor("addons"),
+		statusWGMap:     map[string]*sync.WaitGroup{},
+	}
+
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
 
-	nsInformers := dynamicinformer.NewFilteredDynamicSharedInformerFactory(r.dynClient, time.Minute*30, managedNS, nil)
+	nsInformers := dynamicinformer.NewFilteredDynamicSharedInformerFactory(r.dynClient, time.Minute*30, workflowDeployedNS, nil)
 	wfInf := nsInformers.ForResource(common.WorkflowGVR())
-	bldr := ctrl.NewControllerManagedBy(mgr).
-		For(&addonmgrv1alpha1.Addon{}).
-		// Watch workflows created by addon only in addon-manager-system namespace
-		Watches(&source.Informer{Informer: wfInf.Informer().(cache.Informer)}, &handler.EnqueueRequestForOwner{
-			IsController: true,
-			OwnerType:    &addonmgrv1alpha1.Addon{},
-		})
-
-	generatedInformers = informers.NewSharedInformerFactory(r.generatedClient, time.Minute*30)
-
-	err := mgr.Add(manager.RunnableFunc(func(s <-chan struct{}) error {
-		generatedInformers.Start(s)
-		generatedInformers.WaitForCacheSync(s)
-		nsInformers.Start(s)
-		nsInformers.WaitForCacheSync(s)
-		<-s
-		return nil
-	}))
+	// Watch workflows created by addon only in addon-manager-system namespace
+	if err := c.Watch(&source.Informer{Informer: wfInf.Informer().(cache.Informer)}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &addonmgrv1alpha1.Addon{},
+	}, predicate.NewPredicateFuncs(r.workflowHasMatchingNamespace)); err != nil {
+		return nil, err
+	}
+
+	wfInforms := NewwfInformers(generatedInformers, nsInformers, stopChan)
+	err = mgr.Add(wfInforms)
```

Comment:
This is supposed to be a lambda function, RunnableFunc() that should do some action not return, unless that's changed from the controller-runtime version upgrade... How is passing the informer constructor performing the the `wfInforms.Start()` method?

Reply:
Also, `generatedInformers` was never initialized and you've removed usage for that so I don't think that would need to be passed to the constructor.

Reply:
wsInforms.Start() is a manager.Runnable, in the newer controller-runtime, manager.Add() requires Runnable(context) instead of RunnableFunc().